### PR TITLE
enhance(BDD): implements after suite in sanity package

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <ashishranjan738@gmail.com>

1. This commit implements after suite function in sanity package which is responsible for cleanup of openebs components after the test.
2. And also implements delete functionality in resource.go of `pkg/client/k8s/v1alpha1` which will allow deletion of objects using this package.

-> Testing namespace can be overrided by `--namespace` flag.

output of test:
```
=== RUN   TestSource
Running Suite: Sanity
=====================
Random Seed: 1550226072
Will run 0 of 15 specs

P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:9
    should apply the cstor pvc
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:10
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:9
    should create cstor target
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:14
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:9
    should create cstor volume replica CR
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:18
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:9
    should create cstor replica CR
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:22
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:9
    should create cstor target service
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:26
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should delete cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:31
    should delete cstor target
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:32
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should delete cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:31
    should delete cstor volume replica CR
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:36
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should delete cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:31
    should delete cstor replica CR
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:40
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should delete cstor volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:31
    should delete cstor target service
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:44
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create jiva volumes
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:49
    should create jiva controller
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:50
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create jiva volumes
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:49
    should create jiva replica
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:54
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should create jiva volumes
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:49
    should create jiva controller service
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:58
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should  delete jiva volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:63
    should delete jiva controller
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:64
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should  delete jiva volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:63
    should delete jiva replica
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:68
------------------------------
P [PENDING]
Sanity
/home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:8
  should  delete jiva volume
  /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:63
    should delete jiva controller service
    /home/ashishranjan738/Project/go/src/github.com/openebs/maya/integration-tests/sanity/sanity_test.go:72
------------------------------

Ran 0 of 15 Specs in 40.313 seconds
SUCCESS! -- 0 Passed | 0 Failed | 15 Pending | 0 Skipped
--- PASS: TestSource (40.31s)
PASS
ok      github.com/openebs/maya/integration-tests/sanity        40.324s
```

### CODE REVIEW

File: integration-tests/kubernetes/utils.go [hunk 1]
```diff
@@ -17,3 +20,28 @@ func CheckPodsRunning(pods v1.PodList, expectedPods int) bool {
 	}
 	return true
 }
+
+// CheckForOpenebsNamespace returns true if openebs namespace exists in v1.NamespaceList
+func CheckForOpenebsNamespace(namespaces v1.NamespaceList) bool {
```
Comment (AmitKumarDas) [inline integration-tests/kubernetes/utils.go]: Set a TODO here.
move this to `pkg/namespace/v1alpha1`
This package will have structures, predicates & other functions !!!

File: integration-tests/sanity/sanity_suite_test.go [hunk 2]
```diff
@@ -41,15 +41,15 @@ var _ = BeforeSuite(func() {
...
+	clientset, err := kubernetes.GetClientSet()
 	Expect(err).NotTo(HaveOccurred())
 
-	pods, err := cl.CoreV1().Pods("openebs").List(metav1.ListOptions{})
-	Expect(err).NotTo(HaveOccurred())
-	Expect(pods).NotTo(BeNil())
-
 	status := false
 	for i := 0; i < 100; i++ {
+		pods, err := clientset.CoreV1().Pods("openebs").List(metav1.ListOptions{})
```
Comment (AmitKumarDas) [inline integration-tests/sanity/sanity_suite_test.go]: Get the namespace value as "openebs" as a flag & default the flag to "openebs"

File: integration-tests/sanity/sanity_suite_test.go [hunk 3]
```diff
@@ -62,7 +62,36 @@ var _ = BeforeSuite(func() {
...
+	for i := 0; i < 100; i++ {
+		namespaces, err := clientset.CoreV1().Namespaces().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(namespaces).NotTo(BeNil())
+
+		if kubernetes.CheckForOpenebsNamespace(*namespaces) {
+			status = true
+			break
+		}
+		time.Sleep(10 * time.Second)
```
Comment (AmitKumarDas) [inline integration-tests/sanity/sanity_suite_test.go]: Put the value 10 as a constant

File: integration-tests/kubernetes/utils.go [hunk 4]
```diff
@@ -17,3 +20,28 @@ func CheckPodsRunning(pods v1.PodList, expectedPods int) bool {
...
+	for _, namespace := range namespaces.Items {
+		if namespace.GetName() == "openebs" {
+			return false
+		}
+	}
+	return true
+}
+
+// DeleteResource deletes a resource from kubernetes cluster
+func DeleteResource(groupVersionResource schema.GroupVersionResource, namespace, objectName string) error {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply integration-tests/kubernetes/utils.go]: Currently we don't have any delete support in resource.go

File: integration-tests/kubernetes/utils.go (hunk 1 above)
Comment (ashishranjan738) [reply integration-tests/kubernetes/utils.go]: ok

File: integration-tests/kubernetes/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply integration-tests/kubernetes/utils.go]: @ashishranjan738 I have verified that. Does it not imply we should add delete support in resource.go?

File: integration-tests/sanity/sanity_suite_test.go (hunk 3 above)
Comment (ashishranjan738) [reply integration-tests/sanity/sanity_suite_test.go]: done

File: integration-tests/kubernetes/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply integration-tests/kubernetes/utils.go]: sry, I miss understood. Have added this functionality as discussed

File: integration-tests/sanity/sanity_suite_test.go (hunk 2 above)
Comment (ashishranjan738) [reply integration-tests/sanity/sanity_suite_test.go]: How about making it a constant ?

File: pkg/client/k8s/v1alpha1/resource.go [hunk 5]
```diff
@@ -175,3 +196,31 @@ func (r *createOrUpdate) Apply(obj *unstructured.Unstructured, subresources ...s
...
+}
+
+// Delete is a resource that is suitable to be executed as a Delete operation
+type Delete struct {
+	*resource
+	options ResourceDeleteOptions
+}
+
+// DeleteResource returns a new instance of delete resource
+func DeleteResource(gvr schema.GroupVersionResource, namespace string) *Delete {
```
Comment (ashishranjan738) [inline pkg/client/k8s/v1alpha1/resource.go]: A better name for this?

File: integration-tests/sanity/sanity_suite_test.go (hunk 2 above)
Comment (AmitKumarDas) [reply integration-tests/sanity/sanity_suite_test.go]: There will be cases where we want to test it against different namespaces.

File: pkg/client/k8s/v1alpha1/resource.go (hunk 5 above)
Comment (AmitKumarDas) [reply pkg/client/k8s/v1alpha1/resource.go]: We shall do it later

File: integration-tests/sanity/sanity_suite_test.go (hunk 2 above)
Comment (ashishranjan738) [reply integration-tests/sanity/sanity_suite_test.go]: updated to take it as flag `--namespace`.

**Pull request merged**